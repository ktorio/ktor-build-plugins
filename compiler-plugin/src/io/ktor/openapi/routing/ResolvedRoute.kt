package io.ktor.openapi.routing

data class ResolvedRoute(
    val path: String,
    val method: String,
    val fields: RouteFieldList = emptyList(),
)