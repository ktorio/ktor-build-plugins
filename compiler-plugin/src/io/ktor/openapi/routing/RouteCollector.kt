package io.ktor.openapi.routing

import org.jetbrains.kotlin.fir.FirSession

object RouteCollector {
    fun collectRoutes(graph: RouteCallGraph): List<ResolvedRoute> {
        // Process each route node to generate ResolvedRoutes
        return graph.routes.flatMap { route ->
            if (route.method == null) return@flatMap emptyList()
            
            // Find all paths from roots to this route
            val allPaths = graph.findAllPathsToRoots(route)
            
            // Process each path and merge fields
            with(graph.session) {
                allPaths.mapNotNull { path ->
                    resolve(
                        parents = path,
                        children = graph.getChildren(route)
                    )
                }
            }
        }
    }

    context(session: FirSession)
    fun resolve(
        parents: Iterable<RouteNode> = emptyList(),
        children: Iterable<RouteNode> = emptyList(),
    ): ResolvedRoute? {
        var stack = RouteStack(session)
        var fields = emptyList<RouteField>()
        for (ancestor in parents) {
            fields = ancestor.fields(stack).merge(fields)
            stack += ancestor
        }

        for (child in children) {
            fields = fields.merge(child.fields(stack))
        }

        val path = fields.path ?: return null
        val method = fields.method ?: return null
        if (RouteField.Ignore in fields) return null

        return ResolvedRoute(
            path = path,
            method = method,
            fields = fields,
        )
    }
}