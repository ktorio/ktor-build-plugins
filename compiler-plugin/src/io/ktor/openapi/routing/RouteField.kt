package io.ktor.openapi.routing

import io.ktor.openapi.model.JsonSchema
import io.ktor.openapi.model.JsonType
import kotlinx.serialization.json.JsonElement
import org.jetbrains.kotlin.utils.addToStdlib.firstIsInstanceOrNull

/**
 * Sealed class representing different KDoc parameters for OpenAPI documentation.
 */
sealed interface RouteField {

    fun merge(other: RouteField): RouteField? =
        if (this::class == other::class) this else null

    sealed interface SchemaHolder : RouteField {
        val schema: SchemaReference?
        val attributes: Map<String, JsonElement>
    }

    sealed interface Content : SchemaHolder {
        val contentType: String?
        val description: String?
    }

    sealed interface Parameter : SchemaHolder {
        val name: String
        val description: String?
        val `in`: String
    }

    /**
     * Built into the OpenAPI JSON structure, so ignored when building the path item.
     */
    sealed interface Transient : RouteField

    data class Method(val method: String) : Transient {
        override fun merge(other: RouteField): RouteField? =
            if (other is Method && method == other.method) this else null
    }

    data class Path(val path: String) : Transient {
        override fun merge(other: RouteField): RouteField? =
            if (other is Path)
                Path("${other.path}/${path.removePrefix("/")}")
            else null
    }

    data class Schema(
        val name: String,
        val schema: JsonSchema
    ) : Transient {
        override fun merge(other: RouteField): RouteField? =
            if (other is Schema && other.name == name) this else null
    }

    /**
     * Associates the endpoint with a tag for grouping.
     *
     * Format: `@tag TagName`
     */
    data class Tag(val name: String) : RouteField {
        override fun merge(other: RouteField): RouteField? =
            if (other is Tag && name == other.name) this else null
    }

    /**
     * Describes a path parameter.
     *
     * Format: `@path name description`
     */
    data class PathParam(
        override val name: String,
        override val schema: SchemaReference? = null,
        override val description: String? = null,
        override val attributes: Map<String, JsonElement> = emptyMap(),
    ) : Parameter {
        override val `in`: String = "path"
        override fun merge(other: RouteField): RouteField? =
            if (other is PathParam && name == other.name) copy(
                schema = schema ?: other.schema,
                description = description ?: other.description,
                attributes = attributes + other.attributes,
            ) else null
    }

    /**
     * Describes a query parameter.
     *
     * Format: `@query name description`
     */
    data class QueryParam(
        override val name: String,
        override val schema: SchemaReference? = null,
        override val description: String? = null,
        override val attributes: Map<String, JsonElement> = emptyMap(),
    ) : Parameter {
        override val `in`: String = "query"
        override fun merge(other: RouteField): RouteField? =
            if (other is QueryParam && name == other.name) copy(
                schema = schema ?: other.schema,
                description = description ?: other.description,
                attributes = attributes + other.attributes,
            ) else null
    }

    /**
     * Describes a header parameter.
     *
     * Format: `@header name description`
     */
    data class Header(
        override val name: String,
        override val schema: SchemaReference? = null,
        override val description: String? = null,
        override val attributes: Map<String, JsonElement> = emptyMap(),
    ) : Parameter {
        override val `in`: String = "header"
        override fun merge(other: RouteField): RouteField? =
            if (other is Header && name == other.name) copy(
                schema = schema ?: other.schema,
                description = description ?: other.description,
            ) else null
    }

    /**
     * Describes a cookie parameter.
     *
     * Format: `@cookie name description`
     */
    data class Cookie(
        override val name: String,
        override val schema: SchemaReference? = null,
        override val description: String? = null,
        override val attributes: Map<String, JsonElement> = emptyMap(),
    ) : Parameter {
        override val `in`: String = "cookie"
        override fun merge(other: RouteField): RouteField? =
            if (other is Cookie && name == other.name) copy(
                schema = schema ?: other.schema,
                description = description ?: other.description,
                attributes = attributes + other.attributes,
            ) else null
    }

    /**
     * Documents the request body type.
     *
     * Format: `@body [Type] description`
     */
    data class Body(
        override val contentType: String? = null,
        override val schema: SchemaReference? = null,
        override val description: String? = null,
        override val attributes: Map<String, JsonElement> = emptyMap(),
    ) : Content {
        override fun merge(other: RouteField): RouteField? =
            if (other is Body) copy(
                contentType = contentType ?: other.contentType,
                schema = schema ?: other.schema,
                description = description ?: other.description,
                attributes = attributes + other.attributes,
            ) else null
    }

    /**
     * Documents a response code with optional type and description.
     *
     * Format: `@response code [Type] description`
     */
    data class Response(
        val code: String,
        override val contentType: String? = null,
        override val schema: SchemaReference? = null,
        override val description: String? = null,
        override val attributes: Map<String, JsonElement> = emptyMap(),
    ) : Content {
        override fun merge(other: RouteField): RouteField? =
            if (other is Response && code == other.code)
                copy(
                    contentType = contentType ?: other.contentType,
                    schema = schema ?: other.schema,
                    description = description ?: other.description,
                    attributes = attributes + other.attributes,
                )
            else null
    }

    /**
     * Marks an endpoint as deprecated.
     *
     * Format: `@deprecated reason`
     */
    data class Deprecated(val reason: String) : RouteField

    /**
     * Provides a detailed endpoint description.
     *
     * Format: `@description text`
     */
    data class Description(val text: String) : RouteField


    /**
     * Provides a summary of the endpoint.
     * Format: `@summary text`
     */
    data class Summary(val text: String) : RouteField

    /**
     * Documents security requirements.
     *
     * Format: `@security scheme`
     */
    data class Security(val scheme: String) : RouteField {
        override fun merge(other: RouteField): RouteField? =
            if (other is Security && scheme == other.scheme) this else null
    }
}

typealias RouteFieldList = List<RouteField>

val RouteFieldList.path: String?
    get() = firstIsInstanceOrNull<RouteField.Path>()?.path

val RouteFieldList.method: String?
    get() = firstIsInstanceOrNull<RouteField.Method>()?.method

/**
 * Merges two lists of route fields into a new list.
 */
fun RouteFieldList.merge(other: RouteFieldList) = buildList {
    val otherMutable = other.toMutableList()
    for (field in this@merge) {
        val match = otherMutable.indices.firstNotNullOfOrNull { i ->
            field.merge(otherMutable[i])?.also {
                otherMutable.removeAt(i)
                add(it)
            }
        }
        // if no compatible field was found, add this field to the result
        if (match == null) add(field)
    }
    // add all remaining unmerged fields to the result
    // TODO ignore all summary fields because we currently have some false positives
    addAll(otherMutable.filterNot { it is RouteField.Summary })
}

sealed interface SchemaReference {

    fun asSchema(): JsonSchema

    data class Resolved(
        val schema: JsonSchema
    ) : SchemaReference {
        override fun asSchema(): JsonSchema = schema
    }

    sealed interface Link: SchemaReference {
        val name: String

        data class Simple(override val name: String, val jsonType: JsonType) : Link {
            override fun asSchema(): JsonSchema =
                JsonSchema(jsonType)
        }
        data class Reference(override val name: String) : Link {
            override fun asSchema(): JsonSchema =
                JsonSchema(ref = "#/components/schemas/$name")
        }
        data class Array(val element: Link) : Link by element {
            override fun asSchema(): JsonSchema =
                JsonSchema(
                    type = JsonType.array,
                    items = element.asSchema()
                )
        }
        data class Optional(val delegate: Link) : Link by delegate {
            override fun asSchema(): JsonSchema =
                delegate.asSchema()
                    .copy(required = false)
        }
    }
}

fun SchemaReference.hasReference(): Boolean = getReference() != null

fun SchemaReference.getReference(): String? = when(this) {
    is SchemaReference.Resolved -> schema.ref?.substringAfterLast('/')
    is SchemaReference.Link.Array -> element.getReference()
    is SchemaReference.Link.Optional -> delegate.getReference()
    is SchemaReference.Link.Reference -> name
    is SchemaReference.Link.Simple -> null
}