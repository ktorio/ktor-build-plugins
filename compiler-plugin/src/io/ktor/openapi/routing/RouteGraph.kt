package io.ktor.openapi.routing

import org.jetbrains.kotlin.fir.FirSession

/**
 * Represents a graph of routes where a node can have multiple parents.
 */
class RouteGraph(val session: FirSession) {
    private val parentToChildren = mutableMapOf<RouteNode, MutableSet<RouteNode>>()
    private val childToParents = mutableMapOf<RouteNode, MutableSet<RouteNode>>()
    internal val routes = mutableListOf<RouteNode.Route>()
    internal val functions = mutableListOf<RouteNode.Function>()

    fun isEmpty() = routes.isEmpty()

    fun add(node: RouteNode) {
        for (i in routes.indices.reversed()) {
            val parent = routes[i]
            if (node in parent) {
                addEdge(parent, node)
                break
            } else if (parent.filePath != node.filePath || parent.isTopLevel()) {
                break
            }
        }
        when (node) {
            is RouteNode.Function -> functions.add(node)
            is RouteNode.Route -> routes.add(node)
            is RouteNode.CallFeature -> {}
        }
    }

    /**
     * Now that the function nodes are all accounted for, we can look for their children to
     * populate the edges.
     */
    fun build(): RouteGraph {
        for (function in functions) {
            for (other in functions) {
                if (other in function && other.isTopLevel())
                    addEdge(other, function)
            }
            for (route in routes) {
                if (route in function && route.isTopLevel())
                    addEdge(function, route)
            }
        }
        return this
    }

    fun RouteNode.isTopLevel(): Boolean =
        childToParents[this].orEmpty().none { it is RouteNode.Route }

    /**
     * Adds a parent-child relationship between nodes.
     */
    private fun addEdge(parent: RouteNode, child: RouteNode) {
        parentToChildren.getOrPut(parent) { mutableSetOf() }.add(child)
        childToParents.getOrPut(child) { mutableSetOf() }.add(parent)
    }

    /**
     * Checks if a node has any parents.
     */
    private fun hasParents(node: RouteNode): Boolean =
        childToParents.containsKey(node) && childToParents[node]?.isNotEmpty() == true

    /**
     * Gets all parents of a node.
     */
    private fun getParents(node: RouteNode): Set<RouteNode> =
        childToParents[node] ?: emptySet()

    /**
     * Gets all children of a node.
     */
    fun getChildren(node: RouteNode): Set<RouteNode> =
        parentToChildren[node] ?: emptySet()

    /**
     * Finds all paths from a node to root nodes (nodes with no parents).
     * Returns a list of paths, where each path is a list of nodes from the node to a root.
     */
    fun findAllPathsToRoots(node: RouteNode): List<List<RouteNode>> {
        if (!hasParents(node)) return listOf(listOf(node))

        val result = mutableListOf<List<RouteNode>>()
        val parents = getParents(node)

        for (parent in parents) {
            val parentPaths = findAllPathsToRoots(parent)
            for (path in parentPaths) {
                result.add(path + node)
            }
        }

        return result
    }
}