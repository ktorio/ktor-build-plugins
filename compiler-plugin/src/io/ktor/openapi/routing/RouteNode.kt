package io.ktor.openapi.routing

import io.ktor.compiler.utils.getArgument
import io.ktor.compiler.utils.getFunctionName
import io.ktor.compiler.utils.range
import org.jetbrains.kotlin.fir.expressions.FirFunctionCall
import org.jetbrains.kotlin.fir.expressions.FirLiteralExpression

/**
 * Code references to the Ktor routing API.  These recorded, then merged to form the model of the OpenAPI specification.
 */
sealed interface RouteNode: SourceCoordinates {
    val fir: FirFunctionCall
    val fields: RouteStack.() -> RouteFieldList

    val id: String get() = filePath.orEmpty() + startOffset.toString()
    val functionName: String get() = fir.getFunctionName()
    override val startOffset: Int get() = fir.source?.range?.first ?: -1
    override val endOffset: Int get() = fir.source?.range?.last ?: -1

    /**
     * Route call with method and path.
     *
     * These map to paths in the OpenAPI specification.
     */
    class Route(
        override val filePath: String?,
        override val fir: FirFunctionCall,
        override val fields: RouteStack.() -> RouteFieldList,
    ): RouteNode {
        val method: String? get() =
            functionName.takeIf { it in RoutingFunctionConstants.HTTP_METHODS }

        override fun toString(): String = buildString {
            append(fir.getFunctionName())
            (fir.getArgument("path") as? FirLiteralExpression)?.let {
                append(" ")
                append(it.value.toString())
            }
        }

        override fun equals(other: Any?): Boolean =
            other is Route && id == other.id

        override fun hashCode(): Int = id.hashCode()
    }

    /**
     * A general call to some Ktor feature that generates automatic OpenAPI properties.
     */
    class CallFeature(
        override val filePath: String?,
        override val fir: FirFunctionCall,
        override val fields: RouteStack.() -> RouteFieldList,
    ) : RouteNode {
        override fun equals(other: Any?): Boolean =
            other is CallFeature && id == other.id

        override fun hashCode(): Int = id.hashCode()
    }

    /**
     * Custom extension function of `Route`.
     *
     * Since it's a reference to a function, it may have non-local children.
     */
    class Function(
        override val filePath: String?,
        override val fir: FirFunctionCall,
        val declaration: SourceCoordinates,
        override val fields: RouteStack.() -> RouteFieldList,
    ): RouteNode {
        override fun contains(other: SourceCoordinates): Boolean =
            other in declaration && other != this

        override fun toString(): String =
            fir.getFunctionName() + '@' + fir.source?.range

        override fun equals(other: Any?): Boolean =
            other is Function && id == other.id

        override fun hashCode(): Int = id.hashCode()
    }
}