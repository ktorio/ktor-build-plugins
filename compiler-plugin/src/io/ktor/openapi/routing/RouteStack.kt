package io.ktor.openapi.routing

import io.ktor.compiler.utils.*
import org.jetbrains.kotlin.fir.FirSession
import org.jetbrains.kotlin.fir.expressions.FirExpression

data class RouteStack(
    val session: FirSession,
    val stack: List<RouteNode> = emptyList(),
) {
    operator fun plus(node: RouteNode): RouteStack =
        RouteStack(session, stack + node)

    fun FirExpression.evaluate() =
        with(session) {
            createCallStackEvaluator(stack.map { it.fir })
        }.evaluate(this)
}