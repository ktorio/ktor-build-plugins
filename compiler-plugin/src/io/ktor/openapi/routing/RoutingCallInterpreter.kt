package io.ktor.openapi.routing

import io.ktor.openapi.model.JsonSchema
import kotlinx.serialization.Serializable
import org.jetbrains.kotlin.diagnostics.DiagnosticReporter
import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
import org.jetbrains.kotlin.fir.expressions.FirFunctionCall

/**
 * Infers aspects for OpenAPI generation from a general function call in the source code.
 */
fun interface RoutingCallInterpreter {
    context(context: CheckerContext, reporter: DiagnosticReporter)
    fun check(expression: FirFunctionCall): RoutingReferenceResult
}

sealed interface RoutingReferenceResult {
    /**
     * Nothing found for this function call.
     */
    object None : RoutingReferenceResult

    /**
     * Route node found.
     */
    data class Match(val call: RouteNode) : RoutingReferenceResult

    /**
     * `install(ContentNegotiation)`
     */
    data class ContentType(val contentType: io.ktor.openapi.routing.ContentType): RoutingReferenceResult

    /**
     * `authentication {}`
     */
    data class SecurityScheme(
        val name: String,
        val type: String,
        val scheme: String? = null,
        val bearerFormat: String? = null,
        val openIdConnectUrl: String? = null,
        val flows: Map<String, OauthFlow>? = null,
    ) : RoutingReferenceResult
}

@Serializable
data class OauthFlow(
    val authorizationUrl: String,
    val tokenUrl: String,
    val refreshUrl: String? = null,
    val scopes: Map<String, String>? = null,
)