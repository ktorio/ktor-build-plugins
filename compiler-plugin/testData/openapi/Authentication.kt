// RUN_PIPELINE_TILL: BACKEND

package openapi

import com.auth0.jwt.JWT
import com.auth0.jwt.algorithms.Algorithm
import io.ktor.client.*
import io.ktor.client.engine.apache.*
import io.ktor.http.*
import io.ktor.server.application.*
import io.ktor.server.auth.*
import io.ktor.server.auth.jwt.*
import io.ktor.server.response.*
import io.ktor.server.routing.*
import io.ktor.server.sessions.sessions
import io.ktor.server.sessions.set

enum class UserRole {
    ADMIN,
    USER,
    GUEST
}

data class User(
    val id: String,
    val username: String,
    val roles: Set<UserRole>,
    val scopes: Set<String>
)

class RoleBasedPrincipal(
    val user: User
) {
    fun hasRole(role: UserRole): Boolean = user.roles.contains(role)
    fun hasScope(scope: String): Boolean = user.scopes.contains(scope)
    fun hasAnyRole(vararg roles: UserRole): Boolean = user.roles.any { it in roles }
}

fun Application.installAuthentication() {
    authentication {
        oauth("auth-oauth-google") {
            urlProvider = { "http://localhost:8080/callback" }
            providerLookup = {
                OAuthServerSettings.OAuth2ServerSettings(
                    name = "google",
                    authorizeUrl = "https://accounts.google.com/o/oauth2/auth",
                    accessTokenUrl = "https://accounts.google.com/o/oauth2/token",
                    requestMethod = HttpMethod.Post,
                    clientId = System.getenv("GOOGLE_CLIENT_ID"),
                    clientSecret = System.getenv("GOOGLE_CLIENT_SECRET"),
                    defaultScopes = listOf(
                        "https://www.googleapis.com/auth/userinfo.profile",
                        "https://www.googleapis.com/auth/userinfo.email",
                        "openid"
                    )
                )
            }
            client = HttpClient(Apache)
        }
        oauth("auth-oauth-github") {
            urlProvider = { "http://localhost:8080/github-callback" }
            providerLookup = {
                OAuthServerSettings.OAuth2ServerSettings(
                    name = "github",
                    authorizeUrl = "https://github.com/login/oauth/authorize",
                    accessTokenUrl = "https://github.com/login/oauth/access_token",
                    requestMethod = HttpMethod.Post,
                    clientId = System.getenv("GITHUB_CLIENT_ID"),
                    clientSecret = System.getenv("GITHUB_CLIENT_SECRET"),
                    defaultScopes = listOf("user:email", "repo")
                )
            }
            client = HttpClient(Apache)
        }
        jwt("auth-jwt") {
            realm = "ktor sample app"
            verifier(
                JWT
                    .require(Algorithm.HMAC256("secret"))
                    .withAudience("jwt-audience")
                    .withIssuer("https://jwt-provider-domain/")
                    .build()
            )
            validate { credential ->
                if (credential.payload.audience.contains("jwt-audience")) {
                    val rolesClaim = credential.payload.getClaim("roles")
                    val scopesClaim = credential.payload.getClaim("scope")

                    val roles = rolesClaim.asList(String::class.java)
                        ?.mapNotNull { role -> runCatching { UserRole.valueOf(role) }.getOrNull() }
                        ?.toSet() ?: emptySet()

                    val scopes = scopesClaim.asString()
                        ?.split(" ")
                        ?.toSet() ?: emptySet()

                    val user = User(
                        id = credential.payload.subject,
                        username = credential.payload.getClaim("username").asString() ?: "unknown",
                        roles = roles,
                        scopes = scopes
                    )

                    RoleBasedPrincipal(user)
                } else null
            }
        }
        bearer("auth-api-key-header") {
            realm = "API Access"
            authenticate { tokenCredential ->
                if (tokenCredential.token == "valid-api-key") {
                    UserIdPrincipal("api-user")
                } else null
            }
        }
        basic("auth-api-key-basic") {
            realm = "API Access via Basic Auth"
            validate { credentials ->
                if (credentials.name == "apikey" && credentials.password == "valid-api-key") {
                    UserIdPrincipal("api-user")
                } else null
            }
        }
        basic("auth-cookie-fallback") {
            realm = "Cookie Authentication"
            validate { credentials ->
                if (credentials.name == "session" && credentials.password.isNotEmpty()) {
                    UserIdPrincipal("session-user")
                } else null
            }
        }

    }
}

fun Application.routingWithAuthentication() {
    routing {
        authenticate("auth-oauth-google") {
            get("/login") {
                call.respondRedirect("/callback")
            }

            get("/callback") {
                val principal: OAuthAccessTokenResponse.OAuth2? = call.authentication.principal()
                call.sessions.set(UserSession(principal?.accessToken.toString()))
                call.respondRedirect("/hello")
            }
        }
        authenticate(optional = true) {
            get("/hello") {
                call.respondText("Hello, world!", contentType = ContentType.Text.Plain)
            }
        }
        authenticate("auth-jwt", "auth-api-key-header", "auth-api-key-basic", optional = false) {
            /**
             * Protected endpoint with multiple auth methods
             * @response 200 Authentication successful
             * @response 401 Authentication failed
             */
            get("/protected-multi-auth") {
                call.respondText("Authentication successful")
            }
        }
        authenticate("auth-jwt") {
            /**
             * Admin-only endpoint
             * Requires ADMIN role
             * @response 200 Admin access granted
             * @response 401 Authentication failed
             * @response 403 Authorization failed, requires ADMIN role
             */
            get("/admin") {
                val principal = call.principal<RoleBasedPrincipal>()

                if (principal != null && principal.hasRole(UserRole.ADMIN)) {
                    call.respondText("Admin access granted")
                } else {
                    call.respond(HttpStatusCode.Forbidden, "Requires ADMIN role")
                }
            }

            /**
             * User or admin endpoint
             * Requires USER or ADMIN role
             * @response 200 Access granted
             * @response 401 Authentication failed
             * @response 403 Authorization failed, requires USER or ADMIN role
             */
            get("/user-content") {
                val principal = call.principal<RoleBasedPrincipal>()

                if (principal != null && principal.hasAnyRole(UserRole.USER, UserRole.ADMIN)) {
                    call.respondText("User or admin access granted")
                } else {
                    call.respond(HttpStatusCode.Forbidden, "Requires USER or ADMIN role")
                }
            }
        }
        authenticate("auth-jwt") {
            /**
             * Read-only API endpoint
             * Requires 'api:read' scope
             * @response 200 Read access granted
             * @response 401 Authentication failed
             * @response 403 Authorization failed, requires api:read scope
             */
            get("/api/data") {
                val principal = call.principal<RoleBasedPrincipal>()

                if (principal != null && principal.hasScope("api:read")) {
                    call.respondText("Read access granted")
                } else {
                    call.respond(HttpStatusCode.Forbidden, "Requires api:read scope")
                }
            }

            /**
             * Write API endpoint
             * Requires 'api:write' scope
             * @response 200 Write access granted
             * @response 401 Authentication failed
             * @response 403 Authorization failed, requires api:write scope
             */
            post("/api/data") {
                val principal = call.principal<RoleBasedPrincipal>()

                if (principal != null && principal.hasScope("api:write")) {
                    call.respondText("Write access granted")
                } else {
                    call.respond(HttpStatusCode.Forbidden, "Requires api:write scope")
                }
            }
        }
        route("/conditional") {
            /**
             * Public endpoint for GET, authenticated for POST
             * @response 200 Access granted
             */
            get {
                call.respondText("Public access")
            }

            authenticate("auth-api-key-header") {
                /**
                 * POST requires authentication
                 * @response 200 Authenticated access
                 * @response 401 Authentication failed
                 */
                post {
                    call.respondText("Authenticated access")
                }
            }
        }
        route("/combined-auth") {
            /**
             * Endpoint requiring both JWT and API key
             * @response 200 Combined auth successful
             * @response 401 Authentication failed
             */
            authenticate("auth-jwt") {
                authenticate("auth-api-key-header") {
                    get {
                        call.respondText("Combined authentication successful")
                    }
                }
            }
        }
    }
}

data class UserSession(val accessToken: String)