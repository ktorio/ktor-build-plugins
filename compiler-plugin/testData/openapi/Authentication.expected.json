{
    "openapi": "3.1.1",
    "info": {
        "title": "Open API Document",
        "version": "1.0.0"
    },
    "paths": {
        "/login": {
            "get": {
                "security": [
                    {
                        "auth-oauth-google": []
                    }
                ],
                "responses": {
                    "302": {
                        "description": ""
                    }
                }
            }
        },
        "/callback": {
            "get": {
                "security": [
                    {
                        "auth-oauth-google": []
                    }
                ],
                "responses": {
                    "302": {
                        "description": ""
                    }
                }
            }
        },
        "/hello": {
            "get": {
                "security": [
                    {}
                ],
                "responses": {
                    "200": {
                        "description": "",
                        "content": {
                            "text/plain": {
                                "schema": {
                                    "type": "string"
                                }
                            }
                        }
                    }
                }
            }
        },
        "/protected-multi-auth": {
            "get": {
                "summary": "Protected endpoint with multiple auth methods",
                "responses": {
                    "200": {
                        "description": "Authentication successful",
                        "content": {
                            "text/plain": {
                                "schema": {
                                    "type": "string"
                                }
                            }
                        }
                    },
                    "401": {
                        "description": "Authentication failed"
                    }
                },
                "security": [
                    {
                        "auth-jwt": []
                    },
                    {
                        "auth-api-key-header": []
                    },
                    {
                        "auth-api-key-basic": []
                    }
                ]
            }
        },
        "/admin": {
            "get": {
                "summary": "Admin-only endpoint\nRequires ADMIN role",
                "responses": {
                    "200": {
                        "description": "Admin access granted",
                        "content": {
                            "text/plain": {
                                "schema": {
                                    "type": "string"
                                }
                            }
                        }
                    },
                    "401": {
                        "description": "Authentication failed"
                    },
                    "403": {
                        "description": "Authorization failed, requires ADMIN role",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "type": "string"
                                }
                            }
                        }
                    }
                },
                "security": [
                    {
                        "auth-jwt": []
                    }
                ]
            }
        },
        "/user-content": {
            "get": {
                "summary": "User or admin endpoint\nRequires USER or ADMIN role",
                "responses": {
                    "200": {
                        "description": "Access granted",
                        "content": {
                            "text/plain": {
                                "schema": {
                                    "type": "string"
                                }
                            }
                        }
                    },
                    "401": {
                        "description": "Authentication failed"
                    },
                    "403": {
                        "description": "Authorization failed, requires USER or ADMIN role",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "type": "string"
                                }
                            }
                        }
                    }
                },
                "security": [
                    {
                        "auth-jwt": []
                    }
                ]
            }
        },
        "/api/data": {
            "get": {
                "summary": "Read-only API endpoint\nRequires 'api:read' scope",
                "responses": {
                    "200": {
                        "description": "Read access granted",
                        "content": {
                            "text/plain": {
                                "schema": {
                                    "type": "string"
                                }
                            }
                        }
                    },
                    "401": {
                        "description": "Authentication failed"
                    },
                    "403": {
                        "description": "Authorization failed, requires api:read scope",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "type": "string"
                                }
                            }
                        }
                    }
                },
                "security": [
                    {
                        "auth-jwt": []
                    }
                ]
            },
            "post": {
                "summary": "Write API endpoint\nRequires 'api:write' scope",
                "responses": {
                    "200": {
                        "description": "Write access granted",
                        "content": {
                            "text/plain": {
                                "schema": {
                                    "type": "string"
                                }
                            }
                        }
                    },
                    "401": {
                        "description": "Authentication failed"
                    },
                    "403": {
                        "description": "Authorization failed, requires api:write scope",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "type": "string"
                                }
                            }
                        }
                    }
                },
                "security": [
                    {
                        "auth-jwt": []
                    }
                ]
            }
        },
        "/conditional": {
            "get": {
                "summary": "Public endpoint for GET, authenticated for POST",
                "responses": {
                    "200": {
                        "description": "Access granted",
                        "content": {
                            "text/plain": {
                                "schema": {
                                    "type": "string"
                                }
                            }
                        }
                    }
                }
            },
            "post": {
                "summary": "POST requires authentication",
                "responses": {
                    "200": {
                        "description": "Authenticated access",
                        "content": {
                            "text/plain": {
                                "schema": {
                                    "type": "string"
                                }
                            }
                        }
                    },
                    "401": {
                        "description": "Authentication failed"
                    }
                },
                "security": [
                    {
                        "auth-api-key-header": []
                    }
                ]
            }
        },
        "/combined-auth": {
            "get": {
                "security": [
                    {
                        "auth-api-key-header": []
                    },
                    {
                        "auth-jwt": []
                    }
                ],
                "responses": {
                    "200": {
                        "description": "",
                        "content": {
                            "text/plain": {
                                "schema": {
                                    "type": "string"
                                }
                            }
                        }
                    }
                }
            }
        }
    },
    "components": {
        "schemas": {},
        "securitySchemes": {
            "auth-oauth-google": {
                "type": "oauth2",
                "flows": {
                    "authorizationCode": {
                        "authorizationUrl": "https://accounts.google.com/o/oauth2/auth",
                        "tokenUrl": "https://accounts.google.com/o/oauth2/token",
                        "scopes": {
                            "null": "null"
                        }
                    }
                }
            }
        }
    }
}